pkg/controller/machineset/delete_policy.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+type deletePriority int
+
+const (
+	mustDelete    deletePriority = 100
+	betterDelete  deletePriority = 50
+	couldDelete   deletePriority = 20
+	mustNotDelete deletePriority = 0
+)
+
+type deletePriorityFunc func(machine *v1alpha1.Machine) deletePriority
+
+func simpleDeletePriority(machine *v1alpha1.Machine) deletePriority {
+	if machine.Status.ErrorReason != nil || machine.Status.ErrorMessage != nil {
```

Comment:
The problem with the original issue was order of machines passed to the `getMachinesToDelete` function. The first run of the reconciling loop deleted first machine in the list. The deletion of the machine itself took more time than the reconciling period. So in the second run of the reconciling loop another machine (other than the one that was deleted in the first loop) got deleted. I was able to observe the `List` does not always return the same order of machines. So basically the machineset controller deleted two (sometimes even three) machines instead of one when scaling down by 1.

Even with the improvement the original issue is still reproducible. Looks like the `List` called over CRD resources returns the items in the same order. Still, I would not count on it in the future.

We should take the `deletionTimestamp` into account as well no matter which deletion strategy we use. Any machine that has the timestamp set needs to be returned by `getMachinesToDeletePrioritized` in the top of the list so we don't accidentally delete more than is requested.

Extending the `simpleDeletePriority` with
```go
if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
	return mustDelete
}
```
should do the trick. Would be worth to actually mock the call of the `List` and return the machines in different order.

Reply:
Thanks for your reply in such details, @ingvagabund  I think I have seen the exact problem you described, and it should be addressed. I will add `deletionTimestamp` into account. And the test should be covering the specific issue you mentioned. Great advice!